### PR TITLE
Integrated DataSource elastica driver

### PR DESCRIPTION
**Pull request by chives**

### CODE REVIEW

File: src/Bundle/DataSourceBundle/Elastica/ResultTransformer.php [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+    private array $indexMap = [];
+    private TransformerManager $transformerManager;
+    private ManagerRegistry $managerRegistry;
+
+    public function __construct(
+        ConfigManager $configManager,
+        TransformerManager $transformerManager,
+        ManagerRegistry $managerRegistry
+    ) {
+        $indexNames = $configManager->getIndexNames();
```
Comment (szymach) [inline src/Bundle/DataSourceBundle/Elastica/ResultTransformer.php:39]: Perhaps we could perform this task lazily?

File: src/Component/DataSource/Driver/Elastica/ElasticaDriver.php [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+
+    public function getResult(array $fields, ?int $first, ?int $max): Result
+    {
+        $subQueries = new BoolQuery();
+        $filters = new BoolQuery();
+        $query = $this->masterQuery ?? new Query();
+
+        $this->getEventDispatcher()->dispatch(new PreGetResult($this, $fields, $query));
+
+        if ($this->userFilter !== null) {
```
Comment (szymach) [inline src/Component/DataSource/Driver/Elastica/ElasticaDriver.php:72]: There are a number of places where comparisons are not Yoda style.

File: src/Component/DataSource/Driver/Elastica/ElasticaDriver.php [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+        if ($this->userSubQuery !== null) {
+            $subQueries->addMust($this->userSubQuery);
+        }
+
+        if ($subQueries->hasParam('should') || $subQueries->hasParam('must') || $subQueries->hasParam('must_not')) {
+            $query->setQuery($subQueries);
+        }
+
+        $tempFilters = $filters->getParams();
+        if (!empty($tempFilters)) {
```
Comment (szymach) [inline src/Component/DataSource/Driver/Elastica/ElasticaDriver.php:98]: `0 !== count($tempFilters)`?

File: src/Component/DataSource/Driver/Elastica/FieldType/DateTime.php [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+            );
+        } elseif (in_array($field->getOption('comparison'), ['lt', 'lte', 'gt', 'gte'])) {
+            $filter->addMust(
+                new Range(
+                    $fieldPath,
+                    [$field->getOption('comparison') => $data->format($this->getFormat())]
+                )
+            );
+        } elseif ($field->getOption('comparison') === 'between') {
+            if (!is_array($data)) {
```
Comment (szymach) [inline src/Component/DataSource/Driver/Elastica/FieldType/DateTime.php:50]: This seems to be a common check, we could make an assertion class for it that would throw an exception when needed.

File: src/Component/DataSource/Driver/Elastica/FieldType/DateTime.php [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+                throw new \InvalidArgumentException();
+            }
+
+            if (!empty($data['from'])) {
```
Comment (szymach) [inline src/Component/DataSource/Driver/Elastica/FieldType/DateTime.php:54]: `if (null !== ($data['from'] ?? null))`?

File: src/Bundle/DataSourceBundle/DependencyInjection/FSIDataSourceExtension.php [hunk 6]
```diff
@@ -69,6 +72,9 @@ private function registerDrivers(LoaderInterface $loader): void
         $loader->load('driver/collection.xml');
         $loader->load('driver/doctrine-orm.xml');
         $loader->load('driver/doctrine-dbal.xml');
+        if (true === class_exists(Client::class)) {
```
Comment (szymach) [inline src/Bundle/DataSourceBundle/DependencyInjection/FSIDataSourceExtension.php:75]: Should we also not check for existance of elastica extension?

File: src/Component/DataSource/Driver/Elastica/FieldType/AbstractField.php [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+    }
+
+    public function initOptions(OptionsResolver $optionsResolver): void
+    {
+        parent::initOptions($optionsResolver);
+
+        $optionsResolver
+            ->setDefault('field', null)
+            ->setAllowedTypes('field', ['string', 'null'])
+            ->setNormalizer('field', function (Options $options, $value) {
```
Comment (szymach) [inline src/Component/DataSource/Driver/Elastica/FieldType/AbstractField.php:96]: `static fn`

File: src/Bundle/DataSourceBundle/DependencyInjection/FSIDataSourceExtension.php (hunk 6 above)
Comment (chives) [reply src/Bundle/DataSourceBundle/DependencyInjection/FSIDataSourceExtension.php:75]: `ElasticaDriver` does not require `FOSElasticaBundle` to be loaded if that's what you mean. `FOSElasticaBundle` has to be loaded to use its `ElasticaToModelTransformerInterface` implementing services. That setup is verified and loaded in `FOSElasticaPass`

**Comment by chives** [comment]
@fsi-open

File: src/Component/DataSource/Driver/Elastica/FieldType/Boolean.php [hunk 8]
```diff
@@ -17,12 +17,12 @@
 use FSi\Component\DataSource\Field\Type\BooleanTypeInterface;
 use Symfony\Component\OptionsResolver\OptionsResolver;
 
-class Boolean extends AbstractField implements BooleanTypeInterface
+class Boolean extends AbstractFieldField implements BooleanTypeInterface
```
Comment (szymach) [inline src/Component/DataSource/Driver/Elastica/FieldType/Boolean.php:20]: Is this a mistake?

Comment (chives) [reply src/Component/DataSource/Driver/Elastica/FieldType/Boolean.php:20]: no. this is consistency

**Review by szymach — APPROVED** [review]

**Review by rn0 — APPROVED** [review]

**Pull request merged**